cmd/notation/key.go

```diff
@@ -71,63 +85,104 @@ var (
 )
 
 func addKey(ctx *cli.Context) error {
-	// initialize
-	args := ctx.Args()
-	switch args.Len() {
-	case 0:
-		return errors.New("missing key and certificate paths")
-	case 1:
-		return errors.New("missing certificate path for the correspoding key")
+	cfg, err := config.LoadOrDefault()
+	if err != nil {
+		return err
+	}
+	var key *config.KeySuite
+	pluginName := ctx.String("plugin")
+	if pluginName != "" {
+		key, err = addExternalKey(ctx, pluginName)
+	} else {
+		key, err = newX509KeyPair(ctx)
 	}
-
-	keyPath, err := filepath.Abs(args.Get(0))
 	if err != nil {
 		return err
 	}
-	certPath, err := filepath.Abs(args.Get(1))
+
+	isDefault := ctx.Bool(keyDefaultFlag.Name)
+	err = addKeyCore(cfg, *key, isDefault)
 	if err != nil {
 		return err
 	}
+
+	if err := cfg.Save(); err != nil {
+		return err
+	}
+
+	// write out
+	if isDefault {
+		fmt.Printf("%s: marked as default\n", key.Name)
+	} else {
+		fmt.Println(key.Name)
+	}
+	return nil
+}
+
+func addExternalKey(ctx *cli.Context, pluginName string) (*config.KeySuite, error) {
 	name := ctx.String("name")
 	if name == "" {
-		name = nameFromPath(keyPath)
+		return nil, errors.New("missing key name")
+	}
+	id := ctx.String("id")
+	if id == "" {
+		return nil, errors.New("missing key id")
 	}
+	mgr := manager.NewManager()
+	p, err := mgr.Get(context.Background(), pluginName)
+	if err != nil {
+		return nil, err
+	}
+	if p.Err != nil {
+		return nil, fmt.Errorf("invalid plugin: %w", p.Err)
+	}
+	return &config.KeySuite{
+		Name:        name,
+		ExternalKey: &config.ExternalKey{ID: id, PluginName: pluginName},
+	}, nil
+}
 
-	// check key / cert pair
-	if _, err := tls.LoadX509KeyPair(certPath, keyPath); err != nil {
-		return err
+func newX509KeyPair(ctx *cli.Context) (*config.KeySuite, error) {
+	args := ctx.Args()
+	switch args.Len() {
+	case 0:
+		return nil, errors.New("missing key and certificate paths")
+	case 1:
+		return nil, errors.New("missing certificate path for the corresponding key")
 	}
 
-	// core process
-	cfg, err := config.LoadOrDefault()
+	keyPath, err := filepath.Abs(args.Get(0))
 	if err != nil {
-		return err
+		return nil, err
 	}
-	isDefault, err := addKeyCore(cfg, name, keyPath, certPath, ctx.Bool(keyDefaultFlag.Name))
+	certPath, err := filepath.Abs(args.Get(1))
 	if err != nil {
-		return err
+		return nil, err
 	}
-	if err := cfg.Save(); err != nil {
-		return err
+	name := ctx.String("name")
+	if name == "" {
+		name = nameFromPath(keyPath)
 	}
```

Comment:
Ok, so we derive the name if not explicitly provided. IMO this make discovering the key name to be used difficult when user runs `notation sign --keyName somename` command. How does the user know what to set for `keyName` argument? We should probably remove deriving the name from the path.

Reply:
Agree, this was confusing when I first used the command without specifying the key name.

Reply:
Do we want to make name a required CLI argument then, instead of inferring it, I'd prefer that.

Reply:
Done in [d1bb9be](https://github.com/notaryproject/notation/pull/168/commits/d1bb9be0b6b1fa883bcfe850e2b6568bf8b6ae0e)